src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs

```diff
@@ -114,26 +241,63 @@ private void Initialize(Compilation comp, AnalyzerExecutor analyzerExecutor, Can
             }
         }
 
-        private Task ExecuteSyntaxTreeActions(CancellationToken cancellationToken)
+        private async Task ExecuteSyntaxTreeActions(AnalysisScope analysisScope, AnalysisState analysisStateOpt, CancellationToken cancellationToken)
         {
+            if (analysisScope.IsTreeAnalysis && !analysisScope.IsSyntaxOnlyTreeAnalysis)
+            {
+                // For partial analysis, only execute syntax tree actions if performing syntax analysis.
+                return;
+            }
+
             // Execute syntax tree analyzers in parallel.
-            var tasks = ArrayBuilder<Task>.GetInstance();
-            foreach (var tree in _compilation.SyntaxTrees)
+            var tasks = analysisScope.ConcurrentAnalysis ? ArrayBuilder<Task>.GetInstance() : null;
+
+            try
             {
-                var actionsByAnalyzers = this.analyzerActions.SyntaxTreeActions.GroupBy(action => action.Analyzer);
-                foreach (var analyzerAndActions in actionsByAnalyzers)
+                foreach (var tree in analysisScope.SyntaxTrees)
                 {
-                    var task = Task.Run(() =>
+                    var actionsByAnalyzers = this.analyzerActions.SyntaxTreeActions.ToDictionary(a => a.Analyzer);
+                    foreach (var analyzerAndActions in actionsByAnalyzers)
                     {
+                        var analyzer = analyzerAndActions.Key;
+
                         // Execute actions for a given analyzer sequentially.
-                        analyzerExecutor.ExecuteSyntaxTreeActions(analyzerAndActions.ToImmutableArray(), tree);
-                    }, cancellationToken);
+                        Action executeAnalyzerActions = () =>
```

Comment:
actually one suggestion. I would actually create 2 derived class of TaskRunner base class. one is ConcurrentTaskRunnder and SimpleTaskRunner and then remove the flag and create one of runner at the beginning and run at action through it. each make sure things either run concurrently and serially.